update to 1.12

DerBaM opened the pull request:
My version of an update to 1.12. I know that gudenau allready poste one but he changed a lot unneaded (in my opinion) and that he don't checked his commits (my IDE done this etc.) make this not better.

So my try:
The only big change is the configuration loading. The AbstractConfig.load() would not work (MC crash) so I decided to fix this with "copying" some code from Tinkers Construct istead of fix Mantle to only commit in one place.

In single player tested:
- level up of tools
- level up of weapons
- level up ding and meassage
- config file changes have affect

Fixed also all warnings of my IDE (eclipse) except the deprecated I18n ones.

boosterdev commented:
for server this does not work
java.lang.NoClassDefFoundError: net/minecraft/client/util/ITooltipFlag

DerBaM commented:
Fixed this and the config comments. I've changed them to be clearer but make mistakes.

On server tested:
- level up of tools
- level up of weapons
- level up meassage
- config file changes have affect

Not working on server (for me) is the level up ding.

boosterdev commented:
nice work 👍

KnightMiner commented on src/main/java/slimeknights/toolleveling/EntityXpHandler.java:
```diff
@@ -15,16 +16,18 @@
   @SubscribeEvent
   public void onDeath(LivingDeathEvent event) {
-    if(!event.getEntity().getEntityWorld().isRemote && event.getEntity().hasCapability(CapabilityDamageXp.CAPABILITY, null)) {
+    if (!event.getEntity().getEntityWorld().isRemote && event.getEntity().hasCapability(CapabilityDamageXp.CAPABILITY, null)) {
```
This is minor, but typically we don't have spaces after "if"

DerBaM replied:
Change my IDE settings for this project and fixed all.

KnightMiner commented on src/main/java/slimeknights/toolleveling/Tooltips.java:
```diff
@@ -1,99 +1,49 @@
+
 // utility class for constructing tooltip
-public final class Tooltips {
+@SideOnly(Side.CLIENT)
```
Typically @SideOnly should not be applied to a class as it was never more than a hack. Is there a way to just not reference this class in serverside code like is done with client proxies?

KnightMiner commented on src/main/resources/assets/tinkertoolleveling/lang/en_us.lang:
```diff
@@ -1,31 +1,4 @@
 tooltip.level=Level
 tooltip.xp=XP
 
-tooltip.level.0=Like new
```
Why are these removed? Sure, it does not support every level, but it was much nicer flavor text

DerBaM replied:
It cycled so e.g. clumsy+ is better then awesome. And the color cycle even more:
level * 0.2666 % 1
so there is no info out of the color or the text. Okey, on the other side how much level will you get during playthrough. If this is a deal I have an idea to make us both happy, if you want.

KnightMiner commented on src/main/resources/assets/tinkertoolleveling/lang/de_de.lang:
```diff
@@ -0,0 +1,4 @@
+tooltip.level=Stufe
```
While nice to have, I feel this is better as a separate PR to keep one topic per PR. I'd double check with boni though.

DerBaM replied:
I was in flow (and had time) so I've made my way. But I'm actually all is done what I wanted to finish. So no more comits.

bonii-xx commented on src/main/java/slimeknights/toolleveling/ModToolLeveling.java:
```diff
@@ -93,36 +93,37 @@ public void onBlock(ItemStack tool, EntityPlayer player, LivingHurtEvent event)
+  @SuppressWarnings("unused")
   @SubscribeEvent(priority = EventPriority.LOWEST, receiveCanceled = true)
   public void onLivingHurt(LivingAttackEvent event) {
-    // if it's cancelled it got handled by the battlesign (or something else. but it's a prerequisite.)
```
Those ifs were split up for readability.

DerBaM replied:
There wasn't an emty line between those ifs and that means, that them belong together. And in my opinion it's much more readable now, because you see instand that only one of these conditions must be true.

bonii-xx commented on src/main/java/slimeknights/toolleveling/TinkerToolLeveling.java:
```diff
@@ -9,38 +9,32 @@
+@SuppressWarnings("WeakerAccess")
 @Mod(modid = TinkerToolLeveling.MODID,
-    version = TinkerToolLeveling.VERSION,
-    name = "TinkerToolLeveling",
```
Metadata is missing

bonii-xx commented on src/main/java/slimeknights/toolleveling/TinkerToolLeveling.java:
```diff
@@ -52,20 +46,25 @@ public void preInit(FMLPreInitializationEvent event) {
+  @SuppressWarnings("unused")
   @EventHandler
   public void init(FMLInitializationEvent event) {
+    /* Need this here, because preInit doesn't guarantee that TTL will
```
It does. You removed the dependency that ensures exactly that.

DerBaM replied:
Which dependency you mean? I had the Metadata included with dependencies to mantle/tconstruct and it don't work.

bonii-xx replied:
The one in the @Mod annotation, that you removed. And yes it does work. That's what the whole "after" part in the dependency is about.

DerBaM replied:
No, that @mod stuff wasn't removed on the first commit/push and it doesn't work. I've tested it several times and only after I put this it worked. And I included the missing @mod depeencies in my working brunch (which fixed a few positions you mentioned). But first I've finished the 'clumsy' tooltip stuff and then I look into your mentioned 10 things befor I push it.

bonii-xx commented on src/main/java/slimeknights/toolleveling/CommonProxy.java:
```diff
@@ -4,38 +4,53 @@
+    player.sendMessage(componentText);
+  }
+
+  public static TextFormatting getLevelColor(int level) {
```
No.

bonii-xx replied:
The color is supposed to change every level.

DerBaM replied:
The code was
`float hue = (0.277777f * level);
hue = hue - (int) hue;`
that means that:
lvl  1 - clumsy:  0.277777
lvl 13 - clumsy+: 0,611101
and yes it change every level, but even the level witch the same name had not the same color (or the have all ?143? cycles).

bonii-xx replied:
Same name doesn't require same color. And no, your code only changes color after the first, the 3rd, the 7th etc. level.

DerBaM replied:
Thats right, I've included 10 quality steps. Currently I included the possibility to change the steps and color in the config file.

bonii-xx commented on src/main/java/slimeknights/toolleveling/Tooltips.java:
```diff
@@ -1,99 +1,48 @@
     NBTTagCompound tag = TinkerUtil.getModifierTag(itemStack, TinkerToolLeveling.modToolLeveling.getModifierIdentifier());
     if(!tag.hasNoTags()) {
       ToolLevelNBT data = new ToolLevelNBT(tag);
-      if(Config.canLevelUp(data.level)) {
```
Why would you possibly pull 2 lines into a single line? Even worse, passing a boolean so that a function does A or B is bad, you use 2 functions and call the one that does the logic you need depeding on your condition, not the other way around. Revert to as it was before. It also is more readable that way.

DerBaM replied:
Because to make a one liner: Level 3   XP 3/400
But I wasn't raelly happy about my solution.

bonii-xx commented on src/main/java/slimeknights/toolleveling/ToolLevelNBT.java:
```diff
@@ -1,16 +1,14 @@
-public class ToolLevelNBT extends ModifierNBT {
-
-  private static final String TAG_XP = "xp";
+class ToolLevelNBT extends ModifierNBT {
```
You don't need to make all classes package protected. Just because we don't use them in another package doesn't mean that somebody else wont use them at some point.

DerBaM replied:
"Just because we don't use them in another package" thats a bad reason.

bonii-xx commented on src/main/java/slimeknights/toolleveling/config/Config.java:
```diff
@@ -1,37 +1,138 @@
 
-  public static float getLevelMultiplier() {
-    return INSTANCE.configFile.toolxp.levelMultiplier;
+  public static void syncConfig() {
```
Fix in mantle, if it's broken.

DerBaM replied:
It's now use the style of TC so question back: why TC don't use mantle and fix it but has it's own implementation?

bonii-xx replied:
That is weird indeed. Might be it never got pulled back into Mantle.. or that it requires mod-specific things. What I do know however is that it used to work for TTL, and if it doesn't now then the question is why.

DerBaM replied:
While I change the PR to be step-by-step I can give you the error message (in the file). So you can tell if I make here something wrong or it is truelly a failure of Mantle (I'll wait till I proceed). The game starts anyway but the configuration wouldn't load so that you have default values.

[mantle_config_read_error.txt](https://github.com/SlimeKnights/TinkersToolLeveling/files/1222661/mantle_config_read_error.txt)

bonii-xx replied:
According to that the config file it tries to read is formatted incorrectly. Try deleting it, and if it persists I suggest we simply deactivate the config loading for now (aka only default values) and tackle that separately.

DerBaM replied:
For the above error message but a running not config-loaded game it needs a config file with at least one value. This musn't be one which are needed. In any other cases (no values, emty, no file) the game crash with an other message.
[crash-2017-08-14_18.52.24-client.txt](https://github.com/SlimeKnights/TinkersToolLeveling/files/1222898/crash-2017-08-14_18.52.24-client.txt)

bonii-xx commented on src/main/resources/assets/tinkertoolleveling/lang/en_us.lang:
```diff
@@ -1,31 +1,7 @@
 tooltip.level=Level
 tooltip.xp=XP
 
-tooltip.level.0=Like new
```
No.

bonii-xx commented on src/main/resources/assets/tinkertoolleveling/lang/en_us.lang:
```diff
@@ -1,31 +1,7 @@
-tooltip.level.99=Hacker
+message.levelup=Your %s has reached level %s
 
-message.levelup.2=You begin to feel comfortable handling the %s
```
No.

bonii-xx commented:
>The only big change is the configuration loading. The AbstractConfig.load() would not work (MC crash) so I decided to fix this with "copying" some code from Tinkers Construct istead of fix Mantle to only commit in one place.

No, proper fix is required. Also judging from your code it only broke because you removed all the @mod metadata.

DerBaM commented on build.gradle:
```diff
@@ -3,193 +3,195 @@
  * Master build file
  */
 
+import org.codehaus.groovy.runtime.GStringImpl
```
Personelly I like to remove IDE warnings/hints.

DerBaM commented on build.gradle:
```diff
@@ -3,193 +3,195 @@
+import org.codehaus.groovy.runtime.GStringImpl
+
 buildscript {
-    repositories {
```
Formating change was a seperate push. But this is first time that I use PR. If it make even as seperate push as PR problems I revert it and can it make later as seperate PR. But the much floating/trailing an non formating inner double whitespaces and mix up of tabs and whitespaces should be fixed. I have whitespaces allways activated in my IDE and other editors so I see it. Also make prper formating the source more readable, in my opinion.

bonii-xx replied:
https://blog.jooq.org/2015/03/16/the-10-things-everyone-does-wrong-when-committing-pull-requests/

DerBaM replied:
I'll look at it later and try to make it better.

DerBaM commented:
" because you removed all the @mod metadata" as a seperate LATER commit, right.

bonii-xx commented:
I didn't have time to review it earlier, simple as that. Yes I'm harsh, after all I need to keep maintaining it afterwards too. You wanted to do a port, however most of what you did is not even relevant to a port to 1.12 and made changes on a whim. I'm not saying that the styling is "correct" as is, however formatting does not belong into a PR for a port. If anything it needs a separate PR.

I've put it somewhere above in a comment, but I'm gonna put it here again: https://blog.jooq.org/2015/03/16/the-10-things-everyone-does-wrong-when-committing-pull-requests/

Concentrate on the actual port, and keep the changes only to things that matter to the port. Afterwards there's still time to look at everything else. Every line of formatting changed etc. just adds clutter to the PR.

DerBaM commented:
I'll look at your mentioned blog later and do my best.

DerBaM commented:
bare 1.12 version

It's tested (levelup of tools/weapons, messages, ding sound; on client and server) but configuration is disabled. There diferent error messages depends on if a config file exists or not. Wrote a little more about this above: show outdated Config.java

bonii-xx commented:
Why did you close the PR now?

DerBaM commented:
My failure was, that I don't knew that every push will update the PR. So I had (and do in future) to make seperate branches for PR and personal developement. Sorry for that.

But to contribute on this project the response time is to long for me. That's why I decided to make a fork on which I actually work which is savegamecompatible to this but has it's own lightwight flavor. To not steal you more time and don't confuse other users I closed this and get the repo offline.
